packages/ruleset/src/utils/validate-subschemas.js

```diff
@@ -0,0 +1,36 @@
+const validateSubschemas = (schema, path, validate) => {
+  const errors = [];
+  // invoke validation function
+  errors.push(...validate(schema, path));
+
+  // recursively process subschemas
+  if (schema.properties) {
+    for (const property of Object.entries(schema.properties)) {
+      errors.push(
+        ...validateSubschemas(
+          property[1],
+          [...path, 'properties', property[0]],
+          validate
+        )
+      );
+    }
+  } else if (schema.items) {
+    errors.push(
+      ...validateSubschemas(schema.items, [...path, 'items'], validate)
+    );
+  } else if (schema.allOf || schema.anyOf || schema.oneOf) {
```

Comment:
I know we don't do much with it in our tooling, but is there any reason for us to ignore `not` here? I believe it is completely legal/valid for OAS 3.0.

Reply:
Candidly, I was not aware that existed. I suppose I don't have a good reason to ignore that, I can look into it

Reply:
It is called out in OAS 3.0 (and it's the only other supported "applicator"): https://spec.openapis.org/oas/v3.0.3#schema-object

But reviewing that section of the spec reminded me that we also appear to be missing support for `additionalProperties` schemas. 😁

Reply:
Also... I think almost all these validations should be independent. For example, a schema could have `properties`, and `oneOf` and `not`, and `additionalProperties`.

Of course, there are good reasons for us to prohibit certain combinations (like `anyOf` and `oneOf`) but I kind of thing all these validations (even `items` and `properties`) should be checked independently in this general-purpose function.

Reply:
good catch... I'm embarassed that I didn't catch that myself :)

Reply:
This has now been addressed - `not` and `additionalProperties` are now included and all forms of subschemas are independently validated